libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -121,52 +123,53 @@ string IRGenerationContext::newYulVariable()
-	string funName = IRNames::internalDispatch(_arity);
-	return m_functions.createFunction(funName, [&]() {
-		Whiskers templ(R"(
-			function <functionName>(fun <comma> <in>) <arrow> <out> {
-				switch fun
-				<#cases>
-				case <funID>
-				{
-					<out> <assignment_op> <name>(<in>)
-				}
-				</cases>
-				default { invalid() }
-			}
-		)");
-		templ("functionName", funName);
-		templ("comma", _arity.in > 0 ? "," : "");
-		templ("in", suffixedVariableNameList("in_", 0, _arity.in));
-		templ("arrow", _arity.out > 0 ? "->" : "");
-		templ("assignment_op", _arity.out > 0 ? ":=" : "");
-		templ("out", suffixedVariableNameList("out_", 0, _arity.out));
-
-		vector<map<string, string>> cases;
-		for (FunctionDefinition const* function: collectFunctionsOfArity(_arity))
-		{
-			solAssert(function, "");
-			solAssert(
-				YulArity::fromType(*TypeProvider::function(*function, FunctionType::Kind::Internal)) == _arity,
-				"A single dispatch function can only handle functions of one arity"
-			);
-			solAssert(!function->isConstructor(), "");
-			// 0 is reserved for uninitialized function pointers
-			solAssert(function->id() != 0, "Unexpected function ID: 0");
-
-			cases.emplace_back(map<string, string>{
-				{"funID", to_string(function->id())},
-				{"name", IRNames::function(*function)}
-			});
+void IRGenerationContext::initializeInternalDispatch(InternalDispatchMap _internalDispatch)
+{
+	solAssert(internalDispatchClean(), "");
 
+	for (set<FunctionDefinition const*> const& functions: _internalDispatch | boost::adaptors::map_values)
+		for (auto function: functions)
 			enqueueFunctionForCodeGeneration(*function);
-		}
 
-		templ("cases", move(cases));
-		return templ.render();
-	});
+	m_internalDispatchMap = move(_internalDispatch);
+}
+
+InternalDispatchMap IRGenerationContext::consumeInternalDispatchMap()
+{
+	m_directInternalFunctionCalls.clear();
+
+	InternalDispatchMap internalDispatch = move(m_internalDispatchMap);
+	m_internalDispatchMap.clear();
+	return internalDispatch;
+}
+
+void IRGenerationContext::internalFunctionCalledDirectly(Expression const& _expression)
+{
+	solAssert(m_directInternalFunctionCalls.count(&_expression) == 0, "");
+
+	m_directInternalFunctionCalls.insert(&_expression);
+}
+
+void IRGenerationContext::internalFunctionAccessed(Expression const& _expression)
+{
+	solAssert(IRHelpers::referencedFunctionDeclaration(_expression), "");
+
+	if (m_directInternalFunctionCalls.count(&_expression) == 0)
+	{
+		FunctionDefinition const* function = IRHelpers::referencedFunctionDeclaration(_expression);
+		FunctionType const* functionType = TypeProvider::function(*function, FunctionType::Kind::Internal);
+		solAssert(functionType, "");
+
+		YulArity arity = YulArity::fromType(*functionType);
+		m_internalDispatchMap.try_emplace(arity);
+		m_internalDispatchMap[arity].insert(function);
+		enqueueFunctionForCodeGeneration(*function);
+	}
+}
+
+void IRGenerationContext::internalFunctionCalledViaPointer(YulArity const& _arity)
```

Comment:
In which situations do we need this function and cannot just rely on `internalFunctionCalledDirectly` not being called and then the function being added to the dispatch via `internalFunctionAccessed`?

Reply:
Trying to call something via an uninitialized pointer. In this case `internalFunctionAccessed()` or `internalFunctionCalledDirectly()` never get called for this particular function. If there are no other functions of this arity, the corresponding dispatch function does not get generated.

[calling_uninitialized_function.sol](https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/functionCall/calling_uninitialized_function.sol):
```
Invalid IR generated:
:167:32: Error: Function not found.
let expr_17 := dispatch_internal_in_1_out_1(expr_15_functionIdentifier, _4)
^--------------------------^
:167:17: Error: Variable count mismatch: 1 variables and 0 values.
let expr_17 := dispatch_internal_in_1_out_1(expr_15_functionIdentifier, _4)
^-------------------------------------------------------------------------^
```

Reply:
Ah, now I understand! This basically says "we need a dispatch function for this expression" while the others say "add this function to the switch inside the dispatch function" right?

Reply:
Exactly :)

That's why I initially named these functions referring to what they do (register dispatch, add candidate, etc.). Naming them from the "outside" perspective is valid too but hides details like these.

Reply:
Yeah ok, you are right :)